recipe_scrapers/allrecipes.py

```diff
@@ -40,3 +68,74 @@ def instructions(self):
 
     def ratings(self):
         return self.schema.ratings()
+
+    def nutrients(self):
+        def unhandled(n):
+            logger.warn(f"Unhandled AllRecipes extended nutrient format: {n}")
+
+        # Find extended nutrients
+        ext_nutrs = {}
+        for node in self.soup.select(EXT_NUTRS_SEL):
+            tupl = list(node.stripped_strings)
+            if len(tupl) >= 2:
+                name = tupl[0].strip(":")
+                ext_nutrs[name] = tupl[1]
+                if len(tupl) == 3:
+                    if "%" in tupl[2]:
+                        ext_nutrs[name + "%"] = tupl[2].strip(" %")
+                    else:
+                        unhandled(node)
+            else:
+                unhandled(node)
+
+        # Marry basic and extended nutrients, reporting name clashes
+        nutr = self.schema.nutrients()
+        for name, value in ext_nutrs.items():
+            if name in nutr:
+                logger.warn(EXT_NUTRS_CLASH, name, value, ext_nutrs[name])
+                nutr[f"Ext {name}"] = value
+            else:
+                nutr[name] = value
+
+        return nutr
+
+    def nutrients_unitized(self):
+        unitized = {}
+        for name, value in self.nutrients().items():
+            try:
+                new_value = UNITIZERX.match(value).groups()
+            except AttributeError:
+                new_value = (value, None)
+            if name.endswith("%"):
+                if new_value[1]:
+                    logger.warn(EXT_NUTR_PCT, name, value, str(new_value[1]))
+                else:
+                    new_value = (value, "RDA")
+
+            # Special cases: transfer unit types found in name
+            if not new_value[1]:
+                if "calories" in name:
+                    new_value = (value, "calories")
+            unitized[name] = (float(new_value[0]), new_value[1])
+        return unitized
+
+    def to_dict(self, html=False, unitized=False, skip_attribs=None):
+        obj = {}
+        for attrib_name in DICT_FIELDS:
+            if skip_attribs and attrib_name in skip_attribs:
+                continue
+            attrib = getattr(self.__class__, attrib_name, None)
+            if attrib:
+                if inspect.isfunction(attrib) or inspect.ismethod(attrib):
+                    obj[attrib_name] = attrib(self)
+                else:
+                    obj[attrib_name] = attrib
+            else:
+                logger.warn("Expected attrib not found: %s", attrib_name)
+        if html:
+            obj["html"] = str(self.soup)
+        if unitized:
+            obj["nutrients"] = self.nutrients_unitized()
+        else:
+            obj["nutrients"] = self.nutrients()
```

Comment:
There's something about returning multiple different formats of data from the same `to_dict` method (with multiple boolean control flags) that feels a little non-Pythonic.

What's going to be using the results of the `to_dict` method, and will that be specific to the AllRecipes scraper?

Reply:
> [What's using] the `to_dict` method
It's to prep for serialization, for storage on, or return to, another platform (likely MongoDB, in my case.)

It could possibly fit better on the base class `AbstractScraper`, becoming useful to all subclasses, but I haven't been around long enough to understand how other developers are using `recipe-scrapers` and dealing with the scraped objects, i.e. are they storing or representing instances long-term.

`AbstractScraper`'s semantics seem to be: scrape occurs in base class constructor, parsing done by `SchemaOrg` with resulting regularized instance data attached as `self.schema`, and subclass instance methods providing late customization of instance data.

Typically in OOP, instance data are stored as regular attributes directly on the object, such that objects can just be marshalled using a standard serializer lib, with little or no intervention or preparation necessary.

Sooo... `to_dict` provides the illusion of direct serialization, allowing callers to get a definite set of attributes (and at the very least avoiding `host` which hasn't played nice with me) -- allowing to defer decisions about storing the scraped page's bs4 representation, and breaking up measured values from units.

Is there a preferred method or technique for serializing scraper results, @jayaddison?  I suspect that, as there doesn't appear to be any serialization code, that developers using this project are each implementing their own strategy.  `to_dict` might be useful to more than just myself..

> multiple different formats ... ~non-pythonic

Probably due to not making up my mind about which fields I want to store. I figure the parameters control things that other developers might also want to decide to include or exclude at runtime.

Sometimes it helps to go wide with an idea, gain experience with it, then later narrow the interface to match the uncovered actual need. YMMV

Reply:
Thanks @Kevin-Prichard - yep, there's no encouraged or built-in serialization format in this library at the moment; clients are (implicitly) encouraged to serialize from the abstract scrape results to a concrete format themselves.

Roughly speaking: the library does the hard work of making sense of messy web data, and doesn't have opinions about how that normalized form is then turned into other formats - including all the debate that can follow with that - if it is at all.

That said: finding a way to expose the retrieved HTML to callers could be useful and relatively easy to add to the abstract scraper if that'd help for your use case?

Reply:
> expose the retrieved HTML to callers

The `self.schema` result, or `self.soup` ?

> ..add to the abstract scraper if that'd help for your use case?

Sure, while I'm focused on allrecipes.com for the moment, yea I can see branching out to other sites, especially if my analysis of allrecipes points to increasing breadth.

`to_dict` does no harm, for now, but if your insights about this project indicate another approach to exposing crawl results... in retro, `to_dict`'s `html=False, unitized=False` params could be dropped, as a) bunding raw HTML with discrete normalized data breaks separation of concerns, and b) I have no knowledge that unitization would work equally across subclasses (haven't crawled sites outside allrecipes).

Attaching that method to the subclass was driven by that realization I suppose, but leaving the interface at `to_dict(self, skip_attribs=None)` or even `to_dict(self)` produces a site-agnostic result, but then I lose unitization which while it could be done after-the-fact, was a key aspect of my one-method serialization prep approach. Or, make `to_dict(self)` with `to_dict_unitized(self)` for my purposes, tho again applicability to other sites is unanswered.

I would leave in the `DICT_FIELDS` field inclusion list to avoid having that method gain knowledge about `SchemaOrg`'s scope, unless you can think of a reason to do that (maybe prevent having to separately maintain `DICT_FIELDS` upon top-level changes to `SchemaOrg` attribs, and if that class is easily introspected).

Reply:
About unitization: parsing AllRecipes content into an internal representation (even if there's no public method to access that) where quantities are stored with both amounts and units would be good.

Then the existing `nutrients(...)` method could render the same format of results it currently does (providing backwards-compatibility to people already using `recipe-scrapers`), and we could add a new method -- I'd suggest `nutrient_quantities(...)` -- to return a unitized format.

It would be great if the internal representation could be part of the `AbstractScraper` so that we can re-use it elsewhere too; it's easier for us to accept and integrate pull requests like this if they're going to be applicable for other scrapers too.

(also a heads-up: it could take a few more rounds of review to get to a place where we're all ready with the code - I like the general direction where this is going but I get the feeling I'll have a bunch more feedback to share, including line-level details)